src/librustc_typeck/check/_match.rs

```diff
@@ -605,6 +601,7 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
 
         // Type check the path.
         let pat_ty = self.instantiate_value_path(segments, opt_ty, def, pat.span, pat.id);
+        self.write_ty(pat.id, pat_ty);
```

Comment:
Why pulling `write_ty` out of things?
(I moved this common code into `instantiate_value_path`/`instantiate_type_path` just recently.)